fw/http_frame.c

```diff
@@ -2067,224 +2228,456 @@ tfw_h2_frame_process(TfwConn *c, struct sk_buff *skb, struct sk_buff **next)
+		}
 
-		if (frame_sz + next->len > max_sz)
-			break;
-		/* Don't put different message types into the same frame. */
-		if (skb_tfw_tls_type(next) != tls_type)
-			break;
-		/* Don't agregate skbs with different frame types. */
-		if (skb_tfw_flags(next) & other_flag)
-			break;
-		skb_clear_tfw_flag(next, clear_flag);
-		skb_set_tfw_flags(next, SS_F_HTTP2_FRAME_PREPARED);
-		stream->xmit.nskbs++;
-		frame_sz += next->len;
-		skb_tail = next;
+		BUG_ON(!tls_type);
+		BUG_ON(!skb->len);
+
+		if (skb->len > *len) {
+			split = ss_skb_split(skb, *len);
+			if (!split) {
+				ss_skb_queue_head(&stream->xmit.skb_head, skb);
+				r = -ENOMEM;
+				break;
+			}
+
+			ss_skb_queue_head(&stream->xmit.skb_head, split);
+		}
+		*len -= skb->len;
+		ss_skb_entail(sk, skb, mark, tls_type);
 	}
 
-	return min(max_sz, frame_sz);
+	/*
+	 * We use tls_type and mark from skb_head when we entail data in
+	 * socket write queue. So we should set tls_type and mark for the
+	 * new skb_head.
+	 */
+	if (stream->xmit.skb_head) {
+		skb_set_tfw_tls_type(stream->xmit.skb_head, tls_type);
+		stream->xmit.skb_head->mark = mark;
+	}
+
+	return r;
 }
 
-static int
-tfw_h2_make_frames(struct sock *sk, struct sk_buff *skb, TfwH2Ctx *ctx,
-		   TfwStream *stream, TfwFrameType type, unsigned int mss_now,
-		   unsigned int limit, unsigned int *t_tz)
+static inline int
+tfw_h2_insert_frame_header(TfwH2Ctx *ctx, TfwStream *stream, TfwFrameType type,
+			   unsigned int mss, unsigned long *cwnd_awail,
+			   unsigned long *len)
 {
-	int r = 0;
-	char *data;
-	unsigned char buf[FRAME_HEADER_SIZE];
-	const TfwStr frame_hdr_str = { .data = buf, .len = sizeof(buf)};
 	TfwMsgIter it = {
-		.skb = skb,
-		.skb_head = ((struct sk_buff *)&sk->sk_write_queue),
+		.skb_head = stream->xmit.skb_head,
+		.skb = stream->xmit.skb_head,
 		.frag = -1
 	};
+	unsigned char buf[FRAME_HEADER_SIZE];
+	const TfwStr frame_hdr_str = { .data = buf, .len = sizeof(buf)};
 	TfwFrameHdr frame_hdr = {};
-	unsigned long *len = (type == HTTP2_DATA ?
-		&stream->xmit.b_len : &stream->xmit.h_len);
+	unsigned char tls_type = skb_tfw_tls_type(stream->xmit.skb_head);
+	unsigned int mark = stream->xmit.skb_head->mark;
+	unsigned int length;
+	char *data;
+	int r;
 
-	if (WARN_ON_ONCE(limit <= FRAME_HEADER_SIZE))
-		return -EINVAL;
+	if (type == HTTP2_DATA && skb_headlen(stream->xmit.skb_head)) {
+		if ((r = tfw_http_msg_linear_transform(&it)))
+			return r;
+		stream->xmit.skb_head = it.skb_head;
+	}
 
-	data = tfw_http_iter_set_at_skb(&it, skb, stream->xmit.processed);
-	if (!data)
-		return -E2BIG;
+	data = ss_skb_data_ptr_by_offset(stream->xmit.skb_head, 0);
+	if(unlikely(!data))
+		data = stream->xmit.skb_head->data;
 
-	if (type != HTTP2_HEADERS) {
-		/*
-		 * Insert empty header first, because if some fragments will
-		 * be moved from current skb to the next one, skb length will
-		 * be changed.
-		 */
-		r = tfw_h2_insert_frame_header(sk, skb, stream, mss_now, &it,
-					       &data, &frame_hdr_str, t_tz);
```

Comment:
@krizhanovsky This transformation it's pretty unlikely case. It's possible only when we inserting headers to the skb which already contains `body`. In this case we need move `body` from linear data to paged fragments and place headers before the body.

As @EvgeniiMekhanik since `__tfw_http_msg_linear_transform()` not a static anymore, could you move comment that describes this function from `__tfw_http_msg_expand_from_pool()` where it calls to definition of  `__tfw_http_msg_linear_transform()`.

Reply:
Move this comment to #1411